File: ployglots/mta-ployglot-pkg-figma-api/packages/ts/src/logger.ts
// @ts-nocheck
/**
 * Tiny structured logger with token redaction. Matches the shape used
 * elsewhere in the polyglot stack so downstream logs are uniform.
 */

export type LogLevel = 'trace' | 'debug' | 'info' | 'warn' | 'error' | 'silent';

const LEVEL_RANK: Record<LogLevel, number> = {
  trace: 0,
  debug: 1,
  info: 2,
  warn: 3,
  error: 4,
  silent: 5,
};

export interface Logger {
  trace(msg: string, fields?: Record<string, unknown>): void;
  debug(msg: string, fields?: Record<string, unknown>): void;
  info(msg: string, fields?: Record<string, unknown>): void;
  warn(msg: string, fields?: Record<string, unknown>): void;
  error(msg: string, fields?: Record<string, unknown>): void;
  setLevel(level: LogLevel): void;
}

export interface LoggerOptions {
  level?: LogLevel;
  prefix?: string;
  stream?: { write: (chunk: string) => void };
}

/**
 * Mask a Figma token / password so it does not leak into logs.
 * Shows first 4 + last 4 characters.
 */
export function maskToken(token: string | undefined): string {
  if (token === undefined || token.length === 0) return '<empty>';
  if (token.length <= 8) return '***';
  return `${token.slice(0, 4)}…${token.slice(-4)}`;
}

export function createLogger(options: LoggerOptions = {}): Logger {
  let level: LogLevel = options.level ?? (process.env.LOG_LEVEL as LogLevel) ?? 'info';
  const prefix = options.prefix ?? 'figma-api';
  const stream = options.stream ?? process.stderr;

  const emit = (lvl: LogLevel, msg: string, fields?: Record<string, unknown>) => {
    if (LEVEL_RANK[lvl] < LEVEL_RANK[level]) return;
    const payload = {
      ts: new Date().toISOString(),
      level: lvl,
      logger: prefix,
      msg,
      ...(fields ?? {}),
    };
    stream.write(`${JSON.stringify(payload)}\n`);
  };

  return {
    trace: (m, f) => emit('trace', m, f),
    debug: (m, f) => emit('debug', m, f),
    info: (m, f) => emit('info', m, f),
    warn: (m, f) => emit('warn', m, f),
    error: (m, f) => emit('error', m, f),
    setLevel: (l) => {
      level = l;
    },
  };
}

File: ployglots/mta-ployglot-pkg-figma-api/packages/py/figma_api/_logger.py
"""Tiny structured logger + token-masking helper.

Matches the shape used elsewhere in the polyglot stack so downstream
logs stay uniform across the twin packages.
"""

from __future__ import annotations

import json
import os
import sys
from datetime import UTC, datetime
from typing import Any, Literal, Protocol, TextIO

LogLevel = Literal["trace", "debug", "info", "warn", "error", "silent"]

_RANK: dict[LogLevel, int] = {
    "trace": 0,
    "debug": 1,
    "info": 2,
    "warn": 3,
    "error": 4,
    "silent": 5,
}


class Logger(Protocol):
    def trace(self, msg: str, **fields: Any) -> None: ...
    def debug(self, msg: str, **fields: Any) -> None: ...
    def info(self, msg: str, **fields: Any) -> None: ...
    def warn(self, msg: str, **fields: Any) -> None: ...
    def error(self, msg: str, **fields: Any) -> None: ...
    def set_level(self, level: LogLevel) -> None: ...


def mask_token(token: str | None) -> str:
    """Mask a Figma token so it cannot leak into logs.

    Short tokens → ``***``. Long tokens → ``abcd…wxyz``.
    """
    if token is None or token == "":
        return "<empty>"
    if len(token) <= 8:
        return "***"
    return f"{token[:4]}…{token[-4:]}"


class _StreamLogger:
    def __init__(self, *, level: LogLevel, prefix: str, stream: TextIO) -> None:
        self._level: LogLevel = level
        self._prefix = prefix
        self._stream = stream

    def _emit(self, lvl: LogLevel, msg: str, fields: dict[str, Any]) -> None:
        if _RANK[lvl] < _RANK[self._level]:
            return
        payload: dict[str, Any] = {
            "ts": datetime.now(UTC).isoformat(),
            "level": lvl,
            "logger": self._prefix,
            "msg": msg,
        }
        payload.update(fields)
        self._stream.write(json.dumps(payload) + "\n")

    def trace(self, msg: str, **fields: Any) -> None:
        self._emit("trace", msg, fields)

    def debug(self, msg: str, **fields: Any) -> None:
        self._emit("debug", msg, fields)

    def info(self, msg: str, **fields: Any) -> None:
        self._emit("info", msg, fields)

    def warn(self, msg: str, **fields: Any) -> None:
        self._emit("warn", msg, fields)

    def error(self, msg: str, **fields: Any) -> None:
        self._emit("error", msg, fields)

    def set_level(self, level: LogLevel) -> None:
        self._level = level


def create_logger(
    *,
    level: LogLevel | None = None,
    prefix: str = "figma-api",
    stream: TextIO | None = None,
) -> Logger:
    lvl: LogLevel = level or os.environ.get("LOG_LEVEL", "info").lower()  # type: ignore[assignment]
    if lvl not in _RANK:
        lvl = "info"
    return _StreamLogger(level=lvl, prefix=prefix, stream=stream or sys.stderr)
